codec/codectools/token_producers.go

```diff
@@ -0,0 +1,287 @@
+		tk.Node = n
+		return visitFn(tk)
+	case ipld.ReprKind_List:
+		tk.Kind = TokenKind_ListOpen
+		tk.Length = n.Length()
+		tk.Node = n
+		if err := visitFn(tk); err != nil {
+			return err
+		}
+		litr := n.ListIterator()
+		for !litr.Done() {
+			_, v, err := litr.Next()
+			if err != nil {
+				return err
+			}
+			if err := tokenWalk(tk, v, visitFn); err != nil {
+				return err
+			}
+		}
+		tk.Kind = TokenKind_ListClose
+		tk.Node = n
+		return visitFn(tk)
+	case ipld.ReprKind_Null:
+		tk.Kind = TokenKind_Null
+		return visitFn(tk)
+	case ipld.ReprKind_Bool:
+		tk.Kind = TokenKind_Bool
+		tk.Bool, _ = n.AsBool()
+		return visitFn(tk)
+	case ipld.ReprKind_Int:
+		tk.Kind = TokenKind_Int
+		i, _ := n.AsInt()
+		tk.Int = int64(i) // TODO: upgrade all of ipld to use high precision int consistently
+		return visitFn(tk)
+	case ipld.ReprKind_Float:
+		tk.Kind = TokenKind_Float
+		tk.Float, _ = n.AsFloat()
+		return visitFn(tk)
+	case ipld.ReprKind_String:
+		tk.Kind = TokenKind_String
+		tk.Str, _ = n.AsString()
+		return visitFn(tk)
+	case ipld.ReprKind_Bytes:
+		tk.Kind = TokenKind_Bytes
+		tk.Bytes, _ = n.AsBytes()
+		return visitFn(tk)
+	case ipld.ReprKind_Link:
+		tk.Kind = TokenKind_Link
+		tk.Link, _ = n.AsLink()
+		return visitFn(tk)
+	default:
+		panic(fmt.Errorf("unrecognized node kind (%q?)", n.ReprKind()))
+	}
+	return nil
+}
+
+var TokenWalkSkip = errors.New("token walk: skip")
+
+// --- the stepwise token producer system (more complicated; has a userland stack) is below -->
+
+//
+// A TokenReader can be produced from any ipld.Node using NodeTokenizer.
+// TokenReader are also commonly implemented by codec packages,
+// wherein they're created over a serial data stream and tokenize that stream when pumped.
+//
+// TokenReader implementations are encouraged to yield the same token pointer repeatedly,
+// just varying the contents of the value, in order to avoid unnecessary allocations.
+//
+// TODO: as elegant as this is, it's not able to provide much help for error reporting:
+//  if I hit an error while handling the token, I'd like to be able to ask this thing where it thinks it is,
+//   and include that position info in my error report.
+//  Maybe putting position info directly into the Token struct would solve this satisfactorily?
+//   More comments can be found in the Token definition.
+//
+// TODO: this probably ought to take a budget parameter.
+//  It doesn't make much sense if you're walking in-memory data,
+//  but it's sure relevant if you're parsing serial data and want to pass down info for limiting how big of a string is allowed to be allocated.
+type TokenReader func() (next *Token, err error)
+
+type NodeTokenizer struct {
+	// This structure is designed to be embeddable.  Use Initialize when doing so.
+
+	tk  Token // We embed this to avoid allocations; we'll be repeatedly yielding a pointer to this piece of memory.
+	stk nodeTokenizerStack
+}
+
+func (nt *NodeTokenizer) Initialize(n ipld.Node) {
```

Comment:
why not just have a constructor instead? you seem to pretty much always declare a variable and then initialize it, which is a roundabout way to construct a tokenizer

Reply:
I'm aiming for all of these things to be embeddable.  Initializer patterns seem to work better than constructors for things that are embedded.

One of my major lessons learned from [refmt](https://github.com/polydawn/refmt) is that when you want to compose a bunch of these gadgets -- and you almost always want to pair up at least two things, one to produce and one to consume -- it's very desirable to be able to put them both in one struct, so you can allocate them together in one step (and reset them together in one step, later, too).  Refmt didn't do super well at this: because I didn't design from the start with that goal, I ended up with gadgets that did "just one" allocation when they were created... but by the time you put everything together that was needed to serve a typical end-user purpose, several such "just one" allocations had snuck in, somehow.  So, I'm trying to hew really hard to embeddable gadgets as much as possible from the start, this time around.

I guess it's possible to have constructors that return a value (and *not* the pointer), and assign those returned values over embedded fields, and get the desired effect.

But somehow things just seem to shake out better when you use explicit Init and Reset methods on a pointer receiver when aiming to do embed-friendly stuff, IME.